### PR TITLE
Add ReloadOnChange to KeyPerFile configuration provider

### PR DESCRIPTION
**Pull request by Kahbazi**
Addresses #1085

### CODE REVIEW

File: src/Configuration/Config.KeyPerFile/ref/Microsoft.Extensions.Configuration.KeyPerFile.netcoreapp.cs [hunk 1]
```diff
@@ -24,6 +26,8 @@ public partial class KeyPerFileConfigurationSource : Microsoft.Extensions.Config
         public System.Func<string, bool> IgnoreCondition { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         public string IgnorePrefix { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         public bool Optional { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
+        public int ReloadDelay { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
+        public bool ReloadOnChange { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         public Microsoft.Extensions.Configuration.IConfigurationProvider Build(Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
```
Comment (rynowak) [inline src/Configuration/Config.KeyPerFile/ref/Microsoft.Extensions.Configuration.KeyPerFile.netcoreapp.cs]: Will need API review (even if its the same as what other providers do).

File: src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs [hunk 2]
```diff
@@ -19,6 +19,18 @@ public static class KeyPerFileConfigurationBuilderExtensions
         /// <param name="optional">Whether the directory is optional.</param>
         /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
         public static IConfigurationBuilder AddKeyPerFile(this IConfigurationBuilder builder, string directoryPath, bool optional)
+            => builder.AddKeyPerFile(directoryPath, optional, reloadOnChange: false);
```
Comment (rynowak) [inline src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs]: question - do our other config providers do reloading by default? Looks like JSON does not: https://github.com/aspnet/Extensions/blob/master/src/Configuration/Config.Json/src/JsonConfigurationExtensions.cs#L38

Comment (davidfowl) [reply src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs]: We should be consistent so it should be off by default.

File: src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationProvider.cs [hunk 3]
```diff
@@ -30,12 +48,17 @@ private static string TrimNewLine(string value)
         /// Loads the docker secrets.
```
Comment (Kahbazi) [inline src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationProvider.cs]: Should this summary change to something general?

File: src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs [hunk 4]
```diff
@@ -19,6 +19,18 @@ public static class KeyPerFileConfigurationBuilderExtensions
         /// <param name="optional">Whether the directory is optional.</param>
         /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
         public static IConfigurationBuilder AddKeyPerFile(this IConfigurationBuilder builder, string directoryPath, bool optional)
```
Comment (Kahbazi) [inline src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs]: Should I also add another overload with only directoryPath as input, with optional and reloadOnChange set to false to match other providers?

Comment (davidfowl) [reply src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs]: Yes

File: src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs (hunk 2 above)
Comment (HaoK) [reply src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationBuilderExtensions.cs]: Yeah this is expensive and has some unexpected side effects so generally reload should be off by default unless people explicitly make it enabled

**Comment by pranavkm** [comment]
The API looks pretty identical to the file configuration source, so that looks good. @rynowak \ @HaoK could one of you drive getting the implementation approved and merged?

**Comment by HaoK** [comment]
Sure i can drive this

File: src/Configuration/Config.KeyPerFile/test/KeyPerFileTests.cs [hunk 5]
```diff
@@ -217,6 +217,61 @@ void ReloadLoop()
             Assert.Equal("Foo", options.Text);
         }
 
+        [Fact]
```
Comment (HaoK) [inline src/Configuration/Config.KeyPerFile/test/KeyPerFileTests.cs]: Can you add a test where nothing exists to start, and the directory/files get created later so its the nothing found to something case as opposed to just modification scenarios?

File: src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationProvider.cs (hunk 3 above)
Comment (HaoK) [reply src/Configuration/Config.KeyPerFile/src/KeyPerFileConfigurationProvider.cs]: Sure, loads the configuration values

File: src/Configuration/Config.KeyPerFile/test/KeyPerFileTests.cs (hunk 5 above)
Comment (Kahbazi) [reply src/Configuration/Config.KeyPerFile/test/KeyPerFileTests.cs]: @Haok I added the test.

**Comment by HaoK** [comment]
Looks good, thanks @Kahbazi !

**Pull request merged**